### ISSUE TITLE
Limit resource file change operations to one file operation per resource at a time

### ISSUE THREAD

**Issue by alvacouch** [hydroshare/hydroshare#2122]
@ChristinaB reports that when trying to upload too many things at a time, corruption of the resource is possible on [www](https://www.hydroshare.org.) This is still possible on beta and my test systems for 1.10.0. 

This -- however -- is a symptom of a deeper problem. As people really start using this software, there is a valid use case for uploading 10,000 files at once.  Meanwhile, there are many problems in dealing with this: 

1. (This issue) How to avoid file corruption in a massive series of downloads. 
2. (Issue #2125) How to depict the results of extremely large file downloads. 

Alas, **these issues are related.** 

1. A system that simply tries to download 10,000 things and doesn't allow anyone to do anything else is less usable than one that allows other operations while the download is occurring. 
2. A system that can't manipulate 10,000 files in the UI is undesirable as well. 

@aphelionz and I have discussed this matter in some detail and have a basic plan for dealing with this. It is rather involved, however, and a longer term goal rather than a quick fix. See discussion below, after the discussion of various quick fixes that will make things "better" but not "perfect".

**Comment by alvacouch** [comment]
I have discovered one instance in which this might be due to call-by-reference semantics for JavaScript arrays. I replaced the reference with a copy in PR #2118. This is not necessarily the only problem.

**Comment by alvacouch** [comment]
@dtarb @pkdash @hyi @christinaB I think that the ability of the UI to corrupt files via simultaneous operations (in 1.10.0) is now very limited but not impossible. However, the ability of two different users operating on the same resource to create havoc remains, although more limited than before.

One thing that saves us from more havoc is that: 
1. simple imv commands can be made atomic in iRODS. We can tell `imv` not to overwrite existing things (we haven't done that yet, though)
2. when moving a folder or file, `imv` precedes editing the Django database.
3. Then when users try to move the same file at the same time, the first `imv` will win and the second will lose and fail.   
3. Thus, since single moves always start with `imv`, iRODS always disambiguates race conditions. 

After the proposed changes, these situations work fine: 

scenario | outcome
--------------|--------------
two users try to upload the same single file  | irods upload race won by one file, lost by the other; no inconsistency
two users try to move the same single file to different places | irods mv wins race, other one loses; no inconsistency
two users try to move two directories that are contained within one another | non-deterministic outcome: serialized but not corrupt. Race condition on initial imv prevents second imv. 

But moves of multiple objects into the same folder -- or moves from one folder to multiple places by different users -- can violate user expectations and exhibit unexpected behavior: 

what | result
-------|--------
two users dump multiple files into the same folder with conflicts | result depends upon who wins each conflict race. Neither user gets what they want.  
two users move multiple files out of a folder into others | result depends upon who wins each conflict race. Neither user gets what they want. 
 
Note that in these examples, **Django/iRODS synchronization is not violated.** Instead, users obtain very much unexpected results. E.g., **if two users try to move the same set of files to two places, some will go one place, and some to the other!** The reason for this is that something like this results from all the individual race conditions: 

user 1 request | user 2 request 
------|--------
move file 1 |  
  | move file 1 FAILS
  | move file 2
move file 2 FAILS  |
  | move file 3
move file 3 FAILS | 
 
Etc. 

Thus, in database terms, moves of several objects are **not atomic**, in the sense that it is possible to move some of the objects and not others. 

The regular way to proceed with this is via some kind of **resource locking**, that makes individual moves atomic so that the first user to start a request wins the race and the whole request happens, while the second user loses the race. Note that the observations above make it rather easy to lock a resource to prevent these races, because **only the requests that act on multiple objects at the same time need to be guarded.**

Thus, steps in solving this problem include: 
1. Invoke `imv` with the "no overwrite" option. I've already done this for upload, but not for moving things.
2. Use Django locking to disambiguate and remove conflicts from operations that upload and/or move multiple files. (including UI `paste` and `upload`).

**Comment by alvacouch** [comment]
@ChristinaB Note that non-overlapping concurrent uploads (e.g., uploading different sets of files with no naming conflicts) should work in 1.10.0. Overlapping concurrent uploads have a small chance of failing because there is a small window between checking for existence of a file and writing into it in which a conflict can occur. I could remove this conflict by changing the iRODS command used to do this. Haven't done this yet.

**Comment by hyi** [comment]
@alvacouch Keeping simultaneous operations on the same file of the same resource from different users synchronized to avoid race condition is hard and an external server such as apache zookeeper may be needed in order to resolve this synchronization in a more comprehensive way. When I implemented resource versioning, this synchronization issue must be resolved to avoid broken resource versioning chain when multiple users try to create a new version of the same resource at the same time. What I did was that I added a ```locked_time``` field on BaseResource, and if this field is not None, which means this resource is locked by another user, and prevents other users from doing the same operation at this time until the lock is cleared, and to handle the case where the lock is never cleared due to some unforeseen issues such as network failure, the lock is automatically cleared when the elapsed time since the resource is locked exceeds ```settings.RESOURCE_LOCK_TIMEOUT_SECONDS```. Refer to https://github.com/hydroshare/hydroshare/blob/develop/hs_core/views/__init__.py#L622-L656 for implementation details. This is only implemented for UI, not for REST. Perhaps this locked_time field can be leveraged in some way for keeping other simultaneous file operations synchronized. Since iRODS file operations are called from Django side, if we can keep things sychronized on Django side, we don't have to worry about ```imv``` or other iRODS file operations with overwrite option although I agree we have to use overwrite option with caution, for example, ask users whether they really want to overwrite the file; otherwise, overwrite option probably should not be used.

**Comment by horsburgh** [comment]
@alvacouch - I'm intrigued by this, but also a little confused. What happens if User A moves a file. That change gets queued.  Then, before the file is actually moved, User B deletes the same file. That change is queued. Once operation 1 for User A is complete, the file no longer exists in its original location. So, what does the queued operation for User B (file delete) do if the file no longer exits in the same location?

**Comment by dtarb** [comment]
Can some of the complexity of this be resolved by giving iRODS primacy to be the record of the files that are in the resource.  If multiple actions are being done on a resource at the same time, once the dust settles in iRODS, the UI should show the files that actually ended up in iRODS and Django should align its representations with this, since these are the files that are actually there.  I do not think users should be too surprised by race conditions.  It is like multiple people editing in a Google Doc.  The think the last user does persists, and which is last is a random matter of timing.  If the UI lags the files present, and a user gets an error because of that, it is understandable and to some degree acceptable.  For @horsburgh example above if user B attempts to delete a file that A has already moved, getting an error would be acceptable.

**Comment by ChristinaB** [comment]
I also think it's acceptable to prevent users from doing multiple things that are known to cause problems.  For example, when I am doing something on GogoInflight, the entire screen taken up be a message that prevents me from doing any other operation while reading 'We are sending your data down to earth. This may take a moment. Thanks for your patience'.     Personally, when I am uploading data to HydroShare, I would rather see an enormous screen message like that, so I know exactly what's happening, than to see a spinny wheel.   I think it's great to understand the backend solutions (which I don't, but I'm glad Alva does) but it's OK to use more user messaging to avoid problems when certain fixes may take time.

**Comment by alvacouch** [comment]
@horsburgh If two users race to operate on the same thing, the first one wins and the second one loses. So, if you move a file `foo` somewhere, and then later in the queue `foo` gets deleted, the delete will fail. This is pretty much what you want to have when two people conflict. The first request in the queue always wins. 

@dtarb The software as written does not support what you are asking for above. Django is modified along with iRODS, and the definitive end result is the result of the commands that work without the commands that fail.

**Comment by dtarb** [comment]
@alvacouch I am suggesting a principle that should be used in modifying the software to be a bit more robust.  As I see it, there are files in iRODS that are real physical files.  There are files in Django that are entries in a database that record the names of files.  Where these are in disagreement (as a consequence of a race or something else) it is the actual physical files that are in iRODS that are real, so the resolution should be to update the Django database to be consistent with what is in iRODS.  This would then change what a user sees and is able to act on.  It could be implemented using a strategy that refreshes Django whenever a discrepancy is encountered.

**Comment by alvacouch** [comment]
@dtarb @ChristinaB @pkdash @hyi Primacy of iRODS versus Django doesn't solve anything. There are still very bad things that can happen in either case.

The worst of all is that two users try to unzip zipfiles over one another for the same resource at the same time. Currently this is allowed to happen. The result is pure chaos: one file from the first zipfile, one from the second, etc. There is no predictable outcome from this operation. 

@aphelionz and I have discussed this in detail and propose a much simpler solution. 

1. Queue long running changes to resource files (e.g., dropping 10,000 files onto a resource, unzipping a bag, etc.) into Celery. 
2. Assign multiple celery workers to work in parallel on these tasks, with a policy of only doing one thing on a specific resource at a time. 
3. Thus, the celery queue is limited to one long-running action on a specific resource at a time.  
3. "Lock" the resource against other file changes while there is a queued change for the resource in Celery. This turns off the REST API and UI requests until the long-running request is complete. 
4. When the resource is locked in this manner, the UI displays a banner saying "user x is uploading files" (etc). 
4. Thus, definitively prevent having more than one file operation in the queue for a single resource. 

This prevents a multitude of bad things. And it's fairly easy to accomplish. And it makes sense for users. Users cannot get into the situation @horsburgh talks about, in which they queue something for a result they can't predict. They're forced to wait to queue more actions. 

However, they can always go to another resource and do anything they want while waiting. This makes for a good demo.

**Comment by alvacouch** [comment]
@dtarb @hyi Primacy of iRODS will lead to chaos **without** the above change. So the above proposal is a prerequisite before I will agree to primacy of iRODS as a file existence mechanism.

**Comment by hyi** [comment]
@alvacouch A correction on the concurrency use case you mentioned above. Currently, we intentionally don't have ```-f``` flag in ```ibun``` call for unzip_file operation so that an exception will be raised if two persons are unzipping the same file at the same time since file override is not supported without -f flag but rather an exception is raised. A specific customized message written by @dtarb was raised from server to the client to display the message to the user if this happens. That said, your proposal to queue long-running changes to files sounds good to me.

**Comment by dtarb** [comment]
I agree with the changes @alvacouch is suggesting.  I would restate this as "**limit resource file change operations to one file operation per resource at a time**" as the limit is really that the system cannot undertake more than one operation on a resource, it is not limiting to users.  I do not think the -f flag in ibun that @hyi mentions is quite sufficient as there may still be file conflicts from two separate unzip operations that include a file with the same name.

I think that the locking does not need to apply to all operations.  For example a file download operation that does not change the files in a resource, does not have to block other actions, though there is a question related to a delete, move or rename on a file that is in the process of being downloaded disrupting the download, so there may be merit in at least blocking the change from starting till the download has finished for that single file.  I recognize this is a different level of granularity and the complexity of separate rules that have to operate at the file level and the resource level may be unwarranted.

I think there is still merit in iRODS primacy, and perhaps not even storing metadata for individual content files in Django.  Django just records the location of the contents folder as well as other metadata files, like resourcemetadata.xml, and relies on iRODS listings for any other file operations.  This may have the added value of reducing clutter in the database and simplifying unzip operations that are time consuming.

**Comment by eclark-cuahsi** [comment]
Adding @sblack-usu  as this pertains to HS Core Services. @alvacouch  and @sblack-usu  is this truly a bug or is it an enhancement/improvement?

**Comment by dtarb** [comment]
Folded into large file handling.  No need to track this as an issue here as this is being fixed elsewhere.
